sim-lib/src/lib.rs

```diff
@@ -559,7 +559,21 @@ impl Simulation {
         self.run_data_collection(event_receiver, &mut tasks);
 
         // Get an execution kit per activity that we need to generate and spin up consumers for each source node.
-        let activities = self.activity_executors().await?;
+        let activities = match self.activity_executors().await {
+            Ok(a) => a,
+            Err(e) => {
+                // If we encounter an error while setting up the activity_executors,
+                // we need to shutdown and wait for tasks to finish. We have started background tasks in the
+                // run_data_collection function, so we should shut those down before returning.
+                self.shutdown();
+                while let Some(res) = tasks.join_next().await {
+                    if let Err(e) = res {
+                        log::error!("Task exited with error: {e}.");
+                    }
+                }
```

Comment:
Let's pull this out into a closure to de-dup it a little?

Reply:
Any suggestions on what that closure should look like? It will have to be `async` which is always tricky to me.

Reply:
Async closures are dark magic indeed :') Since we're also using it in `produce_simulation_results` it can probably just be a utility [function](https://github.com/bitcoin-dev-project/sim-ln/files/15149267/wait_fn.patch) instead. 

Closure could have been something like this: 
```
let wait_for_tasks = |mut tasks: JoinSet<()>| async move {
        while let Some(res) = tasks.join_next().await {
            if let Err(e) = res {
                log::error!("Task exited with error: {e}.");
            }
        }
};
```